### DISCUSSION TITLE
Performance for tpcc is sql server in-memory db is lower than the regular db

### DISCUSSION THREAD

**Discussion by bbcc1** [TPC-Council/HammerDB#305] [Q&A] [answered]
The performance is 5x slower comparing the regular db (in SSD raid 10) when run with 1 Virtual user . The performance is similar (but still slower) if I increase virtual users to 64. Is this expected?

**Comment by sm-shaw** [comment, marked as answer]
Issue, converted to discussion Q&A.  A quick test (on an older PC) with results shown as below from HammerDB v4.3 shows that this is not the case - with the in-memory database returning 1.35X higher performance than on-disk for a single Virtual User.  

Note that the organization of in-memory and a regular database is different, and this is covered in the HammerDB documentation: 
See section 3.2 here https://www.hammerdb.com/docs/ch04s03.html
https://www.hammerdb.com/docs/ch04s04.html#d0e1563

You should use SSMS and generate the in-memory reports as described in the documentation to understand what impacts on in-memory performance. In particular, note the hash bucket multiplier and related reports depending on the number of tests you run. 

**On-disk**
![sqlssnip1](https://user-images.githubusercontent.com/38044085/148047969-aa7a136e-87bd-4652-ad33-c9d28047d814.PNG)
![sqlssnip2](https://user-images.githubusercontent.com/38044085/148047981-1d7b37db-ff0f-4ae7-b9d8-59c6404b33d1.PNG)
**In-memory**
![sqlsnip3](https://user-images.githubusercontent.com/38044085/148048039-12e6d3d1-4eab-4e9a-b8a0-81e40c8d1eff.PNG)
![sqlsnip4](https://user-images.githubusercontent.com/38044085/148048048-3548fc95-3eb0-4320-b61a-3538e43c8091.PNG)

**Comment by bbcc1** [comment]
May I know how many warehouses did you configure and what is the hash bucket multiplier ? My test is based on 800 warehouses and the TPM is 20% of your testing. For the on-disk one, did you test in HDD instead of SSD?

**Reply by sm-shaw** [reply]
Yes, 10 warehouses is sufficient for testing 1 virtual user https://www.hammerdb.com/blog/uncategorized/how-many-warehouses-for-the-hammerdb-tpc-c-test.  Also clearly for in-memory you need to allow enough memory for the schema configured.  The disk is an SSD with a Haswell (2013) era CPU.

**Comment by bbcc1** [comment]
Thank you. May I know why I see big performance degradation for in-memory db when I use 800 warehouses (big db). I do not see such performance degradation for the on-disk db.

**Reply by sm-shaw** [reply]
The starting point I would recommend is to view the memory allocation and generate the in-memory report as described here https://www.hammerdb.com/docs/ch04s04.html#d0e1563. 
The key difference is that the in-memory database is implemented with hash indexes https://www.mssqltips.com/sqlservertip/3099/understanding-sql-server-memoryoptimized-tables-hash-indexes/. You should use the performance statistics e.g. starting with activity monitor to find the key wait events.
